### PR TITLE
Add C++ allocator and type traits

### PR DESCRIPTION
**Pull request by thewilsonator**
cc @TurkeyMan 

I need this for dealing with stuff like. 

```C++
template<typename T,
         typename = typename std::enable_if<std::is_base_of<NodeRef, T>::value>::type >
class Array : public NodeRef {
...
}
```

### CODE REVIEW

File: src/core/stdcpp/type_traits.d [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+module core.stdcpp.type_traits;
+
+extern(C++, "std"):
+
+struct integral_constant(T, T Val)
+{
+    enum T value = Val;
+    alias value_type = T;
+    alias type = integral_constant!(value_type, value);
```
Comment (rainers) [inline src/core/stdcpp/type_traits.d]: Can this be different from `typeof(this)`?

**Comment by rainers** [comment]
Is it possible to add a test case in https://github.com/dlang/druntime/tree/master/test/stdcpp, especially for the allocate/deallocate functions?

**Comment by thewilsonator** [comment]
Will do.

File: src/core/stdcpp/type_traits.d (hunk 1 above)
Comment (thewilsonator) [reply src/core/stdcpp/type_traits.d]: No idea, will check.

Comment (thewilsonator) [reply src/core/stdcpp/type_traits.d]: Looks like it always is.

**Comment by thewilsonator** [comment]
Done.

**Comment by thewilsonator** [comment]
```
/home/circleci/druntime/test/stdcpp/src/allocator_test.d:11: undefined reference to `std::allocator<int>::allocate(unsigned long, void const*)'
```

Maybe it's not linking the c++ stdlib? How do we do that?

File: test/stdcpp/win64.mak [hunk 2]
```diff
@@ -5,8 +5,12 @@ MODEL=64
 DRUNTIMELIB=druntime64.lib
 CC=cl
 
-test:
-	"$(CC)" -c /Foarray_cpp.obj test\stdcpp\src\array.cpp /EHsc
-	"$(DMD)" -of=test.exe -m$(MODEL) -conf= -Isrc -defaultlib=$(DRUNTIMELIB) -main -unittest test\stdcpp\src\array_test.d array_cpp.obj
-	test.exe
-	del test.exe test.obj array_cpp.obj
+TESTS= allcator array
```
Comment (Geod24) [inline test/stdcpp/win64.mak]: typo

Comment (thewilsonator) [reply test/stdcpp/win64.mak]: Thanks.

File: src/core/stdcpp/allocator.d [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                enum sub = "AAXPE";
+            else
+                enum sub = "AEXPA";
+            foreach (i; 0 .. name.length - sub.length)
+                if (name[i .. i + sub.length] == sub[])
+                    return name[0 .. i + 3] ~ 'Q' ~ name[i + 4 .. $];
+            assert(false, "substitution string not found!");
+        }
+        pragma(mangle, constHack(deallocate.mangleof))
+        void deallocate(T* ptr, size_t count) nothrow @nogc;
```
Comment (rainers) [inline src/core/stdcpp/allocator.d]: This is interesting: deallocate sets the mangling by retrieving and modifying its own mangling. It seems to works, but looks brittle.

Comment (jacob-carlborg) [reply src/core/stdcpp/allocator.d]: Yeah, I did not expect this to work.

Comment (rainers) [reply src/core/stdcpp/allocator.d]: Actually, in my tests (by default with VS2013) the const hack is wrong. Linking succeeds without it.

File: test/stdcpp/win64.mak [hunk 4]
```diff
@@ -5,8 +5,12 @@ MODEL=64
...
-test:
-	"$(CC)" -c /Foarray_cpp.obj test\stdcpp\src\array.cpp /EHsc
-	"$(DMD)" -of=test.exe -m$(MODEL) -conf= -Isrc -defaultlib=$(DRUNTIMELIB) -main -unittest test\stdcpp\src\array_test.d array_cpp.obj
-	test.exe
-	del test.exe test.obj array_cpp.obj
+TESTS= allocator array
+
+test: $(TESTS)
+
+$(TESTS): %.obj %.d %.cpp
```
[the remark this answers is not in the document]
Comment (thewilsonator) [reply test/stdcpp/win64.mak]: > %-rules are not recognized by DM make.

*laments* Why do we have such a terrible build system?

> The output of C++ compilation should be `/Fo$@_cpp.obj`.

Thanks.

> Stilll causes link errors, though.

I think the C++ standard library needs to be linked in. What is it called on windows?

File: src/core/stdcpp/allocator.d (hunk 3 above)
Comment (thewilsonator) [reply src/core/stdcpp/allocator.d]: Wonderful! Do we have a version or `__tratits(getTargetInfo, "blah")` to detect what version of VS we are linking against?

File: test/stdcpp/win64.mak (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacob-carlborg) [reply test/stdcpp/win64.mak]: We do have `build.d` as an alternative for make.

File: src/core/stdcpp/allocator.d (hunk 3 above)
Comment (rainers) [reply src/core/stdcpp/allocator.d]: I retried with VS2015 and VS2017, these also pass without complaining about `deallocate` (without the const hack), but a missing generic delete in the array-test:
```
"c:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023/bin/Hostx64/x64/cl.exe" -c /Foallocator_c
pp.obj test\stdcpp\src\allocator.cpp /EHsc
Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27023.1 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

allocator.cpp

"c:\s\d\rainers\windows\bin\dmd.exe" -of=allocator.exe -m64 -conf= -Isrc -defaultlib=lib\druntime64.lib -main -unittest test\stdcp
p\src\allocator_test.d allocator_cpp.obj

allocator.exe

del allocator.exe allocator.obj allocator_cpp.obj
"c:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023/bin/Hostx64/x64/cl.exe" -c /Foarray_cpp.o
bj test\stdcpp\src\array.cpp /EHsc
Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27023.1 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

array.cpp

"c:\s\d\rainers\windows\bin\dmd.exe" -of=array.exe -m64 -conf= -Isrc -defaultlib=lib\druntime64.lib -main -unittest test\stdcpp\sr
c\array_test.d array_cpp.obj
array_cpp.obj : error LNK2019: unresolved external symbol "void __cdecl operator delete(void *,unsigned __int64)" (??3@YAXPEAX_K@Z
) referenced in function "public: virtual void * __cdecl std::_Generic_error_category::`scalar deleting destructor'(unsigned int)"
 (??_G_Generic_error_category@std@@UEAAPEAXI@Z)
array.exe : fatal error LNK1120: 1 unresolved externals
Error: linker exited with status 1120

--- errorlevel 1
```

Comment (TurkeyMan) [reply src/core/stdcpp/allocator.d]: Yes this line is sensitive to VS version. I wrote this against VS2017, which is 'current', but we may need to be able to specify the target VS version (similar problem to the recent `-cppstd` debate.

**Comment by TurkeyMan** [comment]
Sorry, I've been AFK since before Christmas.

So what's going on here... you've extracted `allocator` and `type_traits` from my `std::string` PR into a distinct PR?
I'm not sure I understand what this is for?

**Comment by thewilsonator** [comment]
@TurkeyMan this is self contained, easier to review, necessary for more things than just `std::string`, e.g. `std::vector` and dealing with `enable_if` BS.

**Comment by Geod24** [comment]
👍 for the allocator split, I agree getting digestable diffs will increase review/merge speed.

Not so found of the `type_traits` part, as from my experience in C++ binding, type traits were always much better dealt with using D's superior reflection abilities.
On the plus side it might allow to keep the code closer to the C++ one, but I think most if not all usage can be replaced with a trivial D construct. Do you have an example of where it shows its true usefulness ?

**Comment by Geod24** [comment]
Oh, that makes perfect sense now. Thanks.

**Comment by thewilsonator** [comment]
> Incidentally, here's another case of re-attributing all the commit metadata.

I ain't got the git-fu for that.

> I might further split this into 2 separate PR's, and we can get the type_traits one in eagerly.

Sure, sounds good.

**Comment by jacob-carlborg** [comment]
> I ain't got the git-fu for that.

A commit can have a separate author and a separate committer. You can change the author using:

```
git commit --amend --author="Author Name <email@address.com>"
```

**Comment by wilzbach** [comment]
> I ain't got the git-fu for that.

It's not that hard. It's actually more effort to misattribute it. If you checkout a PR branch locally, you have the commit(s) from the original author. You might want to squash them before you amend work, but why would you change the metadata. Are you not using git to checkout a PR, but downloading the files individually from GitHub's web interface?

**Comment by TurkeyMan** [comment]
See #2439, and we'll work out the allocator stuff further... it depends on some other things in the works to define the extern's properly.

**Comment by TurkeyMan** [comment]
#2444 should conclude the work in this patch. This should be closed.

**Pull request closed without merging**